File: apps/api/src/routes/voice.ts
import { Router, type Request, type Response } from 'express';
import multer from "multer";
import { interpretMessage } from "../services/ollama.js";
import { executeIntent } from "./utils/executeIntent.js";

const AI_SERVICE_URL = process.env.AI_SERVICE_URL || "http://ai:8000";

// Configure multer for memory storage (audio files)
const upload = multer({
  storage: multer.memoryStorage(),
  limits: {
    fileSize: 10 * 1024 * 1024, // 10MB max
  },
});

export function createVoiceRouter(): Router {
  const router = Router();

  // Transcribe audio to text
  router.post("/transcribe", upload.single("audio"), async (req: Request, res: Response) => {
    try {
      if (!req.file) {
        res.status(400).json({ ok: false, error: "No audio file provided" });
        return;
      }

      const formData = new FormData();
      formData.append("audio", new Blob([new Uint8Array(req.file.buffer)]), req.file.originalname || "audio.wav");

      const response = await fetch(`${AI_SERVICE_URL}/voice/transcribe`, {
        method: "POST",
        body: formData,
      });

      if (!response.ok) {
        const error = await response.text();
        res.status(response.status).json({ ok: false, error });
        return;
      }

      const data = await response.json();
      res.json({ ok: true, ...data });
    } catch (err) {
      console.error("Error transcribing audio:", err);
      res.status(503).json({
        ok: false,
        error: "Voice service unavailable",
      });
    }
  });

  // Synthesize text to speech
  router.post("/synthesize", async (req: Request, res: Response) => {
    try {
      const { message } = req.body;

      if (!message || typeof message !== "string") {
        res.status(400).json({ ok: false, error: "message is required" });
        return;
      }

      const response = await fetch(`${AI_SERVICE_URL}/voice/synthesize`, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({ message }),
      });

      if (!response.ok) {
        const error = await response.text();
        res.status(response.status).json({ ok: false, error });
        return;
      }

      // Stream the audio response
      res.setHeader("Content-Type", "audio/wav");
      const arrayBuffer = await response.arrayBuffer();
      res.send(Buffer.from(arrayBuffer));
    } catch (err) {
      console.error("Error synthesizing speech:", err);
      res.status(503).json({
        ok: false,
        error: "Voice service unavailable",
      });
    }
  });

  // Process voice command end-to-end (audio in, JSON out)
  // Uses AI service for transcription, then API's orchestrator for intent processing
  router.post("/command", upload.single("audio"), async (req: Request, res: Response) => {
    try {
      if (!req.file) {
        res.status(400).json({ ok: false, error: "No audio file provided" });
        return;
      }

      // Step 1: Transcribe using AI service
      const formData = new FormData();
      formData.append("audio", new Blob([new Uint8Array(req.file.buffer)]), req.file.originalname || "audio.wav");

      const transcribeResponse = await fetch(`${AI_SERVICE_URL}/voice/transcribe`, {
        method: "POST",
        body: formData,
      });

      if (!transcribeResponse.ok) {
        const error = await transcribeResponse.text();
        res.status(transcribeResponse.status).json({ ok: false, error });
        return;
      }

      const transcription = (await transcribeResponse.json()) as { text: string; success: boolean };

      if (!transcription.text) {
        res.json({
          ok: true,
          transcription: "",
          response: "I didn't catch that. Could you please repeat?",
        });
        return;
      }

      // Step 2: Process with API's orchestrator (same as chat)
      const intent = await interpretMessage(transcription.text);

      // Step 3: Execute the intent (shared with chat routes)
      const { deviceId, location } = req.body;
      const result = await executeIntent(intent, {
        deviceId,
        location,
        source: "voice",
        message: transcription.text,
      });

      res.json({
        ok: result.ok,
        transcription: transcription.text,
        response: result.reply,
        action: result.action?.type,
        target: result.action?.target as string | undefined,
        value: result.action?.value,
      });
    } catch (err) {
      console.error("Error processing voice command:", err);
      res.status(503).json({
        ok: false,
        error: "Voice service unavailable",
      });
    }
  });

  // Health check for voice service
  router.get("/health", async (_req: Request, res: Response) => {
    try {
      const response = await fetch(`${AI_SERVICE_URL}/health`);
      if (!response.ok) {
        res.json({ ok: false, service: "voice" });
        return;
      }
      const data = await response.json();
      res.json({ ok: true, service: "voice", ...data });
    } catch (err) {
      res.json({ ok: false, service: "voice", error: "Service unavailable" });
    }
  });

  return router;
}

File: apps/cortex/src/api/voice.py
"""
Voice routes: POST /api/voice/transcribe, /api/voice/synthesize, /api/voice/command

These are now direct endpoints (no proxy layer) since voice service lives in the same process.
"""

import logging

from fastapi import APIRouter, File, Request, UploadFile
from fastapi.responses import Response
from pydantic import BaseModel

from ..services.intent_executor import execute_intent

logger = logging.getLogger(__name__)


class SynthesizeRequest(BaseModel):
    message: str


def create_voice_router(sqlite, redis_client, ws_server, ollama, voice_service) -> APIRouter:
    r = APIRouter()

    @r.post("/transcribe")
    async def transcribe_audio(audio: UploadFile = File(...)):
        if not voice_service or not voice_service.is_stt_available():
            return {"ok": False, "error": "Speech-to-text service not available"}

        try:
            audio_data = await audio.read()
            text = voice_service.transcribe(audio_data)
            return {"ok": True, "text": text, "success": True}
        except Exception as e:
            logger.error(f"Transcription failed: {e}")
            return {"ok": False, "error": str(e)}

    @r.post("/synthesize")
    async def synthesize_speech(request: SynthesizeRequest):
        if not voice_service or not voice_service.is_tts_available():
            return {"ok": False, "error": "Text-to-speech service not available"}

        try:
            audio_data = voice_service.synthesize(request.message)
            return Response(content=audio_data, media_type="audio/wav")
        except Exception as e:
            logger.error(f"Synthesis failed: {e}")
            return {"ok": False, "error": str(e)}

    @r.post("/command")
    async def voice_command(audio: UploadFile = File(...), request: Request = None):
        """Full STT → LLM → executeIntent pipeline."""
        try:
            if not voice_service or not voice_service.is_stt_available():
                return {"ok": False, "error": "Voice service unavailable"}

            # Step 1: Transcribe
            audio_data = await audio.read()
            text = voice_service.transcribe(audio_data)

            if not text:
                return {
                    "ok": True,
                    "transcription": "",
                    "response": "I didn't catch that. Could you please repeat?",
                }

            # Step 2: Interpret with LLM
            intent = ollama.interpret_message(text, sqlite, ws_server)

            # Step 3: Execute
            mqtt = getattr(request.app.state, "mqtt", None) if request else None
            result = await execute_intent(
                intent, source="voice", message=text,
                sqlite=sqlite, redis=redis_client, mqtt=mqtt, ws=ws_server,
            )

            return {
                "ok": result["ok"],
                "transcription": text,
                "response": result["reply"],
                "action": result.get("action", {}).get("type"),
                "target": result.get("action", {}).get("target"),
                "value": result.get("action", {}).get("value"),
            }

        except Exception as e:
            logger.error(f"Error processing voice command: {e}")
            return {"ok": False, "error": "Voice service unavailable"}

    @r.get("/health")
    async def voice_health():
        return {
            "ok": True,
            "service": "voice",
            "stt_available": voice_service.is_stt_available() if voice_service else False,
            "tts_available": voice_service.is_tts_available() if voice_service else False,
            "llm_available": ollama.is_available() if ollama else False,
        }

    return r
